pkg/agent/controller/networkpolicy/fqdn.go

```diff
@@ -0,0 +1,782 @@
+func (f *fqdnController) getIPsForFQDNSelectors(fqdns []string) []net.IP {
+	f.fqdnSelectorMutex.Lock()
+	defer f.fqdnSelectorMutex.Unlock()
+	var matchedIPs []net.IP
+	for _, fqdn := range fqdns {
+		fqdnSelectorItem := fqdnToSelectorItem(fqdn)
+		fqdnsMatched, ok := f.selectorItemToFQDN[fqdnSelectorItem]
+		if !ok {
+			klog.Errorf("FQDN selector %v is not known to the controller, cannot get IPs", fqdnSelectorItem)
+			return matchedIPs
+		}
+		for fqdn := range fqdnsMatched {
+			if dnsMeta, ok := f.dnsEntryCache[fqdn]; ok {
+				for _, ip := range dnsMeta.responseIPs {
+					matchedIPs = append(matchedIPs, ip)
+				}
+			}
+		}
+	}
+	return matchedIPs
+}
+
+// addFQDNRule adds a new FQDN rule to fqdnSelectorItem mapping, as well as the OFAddresses of
+// Pods selected by the FQDN rule.
+func (f *fqdnController) addFQDNRule(ruleID string, fqdns []string, podOFAddrs sets.Int32) error {
+	f.addFQDNSelector(ruleID, fqdns)
+	return f.updateRuleSelectedPods(ruleID, podOFAddrs)
+}
+
+func (f *fqdnController) addFQDNSelector(ruleID string, fqdns []string) {
+	f.fqdnSelectorMutex.Lock()
+	defer f.fqdnSelectorMutex.Unlock()
+	for _, fqdn := range fqdns {
+		fqdnSelectorItem := fqdnToSelectorItem(fqdn)
+		ruleIDs, exists := f.selectorItemToRuleIDs[fqdnSelectorItem]
+		if !exists {
+			// This is a new fqdnSelectorItem. All existing FQDNs in the cache needs to be matched
+			// against this fqdnSelectorItem to update the mapping.
+			f.selectorItemToRuleIDs[fqdnSelectorItem] = sets.NewString(ruleID)
+			for fqdn := range f.dnsEntryCache {
+				if fqdnSelectorItem.matches(fqdn) {
+					f.setFQDNMatchSelector(fqdn, fqdnSelectorItem)
+				}
+			}
+		} else {
+			f.selectorItemToRuleIDs[fqdnSelectorItem] = ruleIDs.Insert(ruleID)
+		}
+		if fqdnSelectorItem.matchName != "" {
+			// Start a DNS query immediately for matchName selectors.
+			f.setFQDNMatchSelector(fqdnSelectorItem.matchName, fqdnSelectorItem)
+			f.dnsQueryQueue.Add(fqdnSelectorItem.matchName)
+		}
+	}
+}
+
+// updateRuleSelectedPods updates the Pod OFAddresses selected by a FQDN rule. Those addresses
+// are used to create DNS response interception rules.
+func (f *fqdnController) updateRuleSelectedPods(ruleID string, podOFAddrs sets.Int32) error {
+	f.fqdnRuleToPodsMutex.Lock()
+	defer f.fqdnRuleToPodsMutex.Unlock()
+	originalPodSet, newPodSet := sets.Int32{}, sets.Int32{}
+	for _, pods := range f.fqdnRuleToSelectedPods {
+		originalPodSet = originalPodSet.Union(pods)
+	}
+	f.fqdnRuleToSelectedPods[ruleID] = podOFAddrs
+	for _, pods := range f.fqdnRuleToSelectedPods {
+		newPodSet = newPodSet.Union(pods)
+	}
+	addedPods, removedPods := newPodSet.Difference(originalPodSet), originalPodSet.Difference(newPodSet)
+	if len(addedPods) > 0 {
+		var addedOFAddrs []types.Address
+		for port := range addedPods {
+			addedOFAddrs = append(addedOFAddrs, openflow.NewOFPortAddress(port))
+		}
+		if err := f.ofClient.AddAddressToDNSConjunction(dnsInterceptRuleID, addedOFAddrs); err != nil {
```

Comment:
I took a look at that and where this function is called from. `addFQDNRule` is called from `computeOFRulesForAdd`, which is a bit surprising to me because based on the name of the function and from previous usage, I would expect that this function compute things and doesn't actually install flows...

`computeOFRulesForAdd` already has this TODO:
> // TODO: Handle the case that the following processing fails or partially succeeds.

I don't have good knowledge of this code but here is what I did:
- log error returned by `addFQDNRule` instead of ignoring it
- add a new TODO / comment at the `addFQDNRule` call site